### PR TITLE
Add skippedPropagatingAPIs option

### PR DESCRIPTION
**Pull request by pigletfly**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes https://github.com/karmada-io/karmada/issues/345 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: pkg/util/detector/detector.go [hunk 1]
```diff
@@ -221,13 +227,22 @@ func (d *ResourceDetector) EventFilter(obj interface{}) bool {
...
+		return false
+	}
+	watchedKinds := strings.Split(d.WatchedKinds, ",")
+	if len(watchedKinds) > 0 {
+		kindsSet := sets.NewString(watchedKinds...)
+		if kindsSet.Has(clusterWideKey.Kind) {
+			return true
+		}
+	}
 	return true
```
Comment (XiShanYongYe-Chang) [inline pkg/util/detector/detector.go:246]: mathed resource kind will return true with line 239-245, here return false?

Comment (pigletfly) [reply pkg/util/detector/detector.go:244]: `IgnoreKinds ` such as Node and Event will ignored by default, if watchedKinds is set, then detector will only filter these kinds events.

File: cmd/controller-manager/app/options/options.go [hunk 2]
```diff
@@ -79,4 +83,6 @@ func (o *Options) AddFlags(flags *pflag.FlagSet) {
 	flags.DurationVar(&o.ClusterStatusUpdateFrequency.Duration, "cluster-status-update-frequency", 10*time.Second,
 		"Specifies how often karmada-controller-manager posts cluster status to karmada-apiserver.")
 	flags.BoolVar(&o.LeaderElection.LeaderElect, "leader-elect", true, "Start a leader election client and gain leadership before executing the main loop. Enable this when running replicated components for high availability.")
+	flags.StringVar(&o.WatchedKinds, "watchedKinds", "", "Indicates which API objects are meaningful to karmada-controller.")
```
Comment (mrlihanbo) [inline cmd/controller-manager/app/options/options.go:86]: Multi resource are split by ",", right? You can give an example here, such as "pod, service, deployment"

**Comment by mrlihanbo** [comment]
you need to update the deploy file of karmada controller to make e2e success.

File: cmd/controller-manager/app/options/options.go (hunk 2 above)
Comment (pigletfly) [reply cmd/controller-manager/app/options/options.go:86]: yes, it's comma-delimited

**Comment by mrlihanbo** [comment]
> you need to update the deploy file of karmada controller to make e2e success.

If watchedKinds and WatchedNamespaces are empty, the detector will watch all resources except some reserved namespace and crds. No need to update deploy file of karmada controller.

File: pkg/util/detector/detector.go [hunk 3]
```diff
@@ -221,13 +229,26 @@ func (d *ResourceDetector) EventFilter(obj interface{}) bool {
...
+	if names.IgnoreKinds.Has(clusterWideKey.Kind) {
+		return false
+	}
+	watchedKinds := strings.Split(d.WatchedKinds, ",")
+	if len(watchedKinds) > 0 {
+		kindsSet := sets.NewString(watchedKinds...)
+		return kindsSet.Has(clusterWideKey.Kind)
+	}
+
+	watchedNamespaces := strings.Split(d.WatchedNamespaces, ",")
```
Comment (mrlihanbo) [inline pkg/util/detector/detector.go:247]: It seems like logic here won't be executed if watchedKinds is set?

Comment (pigletfly) [reply pkg/util/detector/detector.go]: good catch.If both watchedKinds and watchedNamespaces, it should only watch watchedKinds in watchedNamespaces.

**Comment by RainbowMango** [comment]
Hi @pigletfly Are you still working on this?
We can get the #345 on board with this PR.

**Comment by pigletfly** [comment]
> Hi @pigletfly Are you still working on this?
> We can get the #345 on board with this PR.

Yes, trying to complete it this week

File: cmd/controller-manager/app/options/options.go [hunk 4]
```diff
@@ -87,4 +89,8 @@ func (o *Options) AddFlags(flags *pflag.FlagSet) {
 		"Specifies the expiration period of a cluster lease.")
 	flags.Float64Var(&o.ClusterLeaseRenewIntervalFraction, "cluster-lease-renew-interval-fraction", 0.25,
 		"Specifies the cluster lease renew interval fraction.")
+	flags.StringVar(&o.SkippedPropagatingAPIs, "skippedPropagatingAPIs", "", "Comma separated resources that should be skipped from propagating in addition to the default skip list(cluster.karmada.io;policy.karmada.io;work.karmada.io). Supported formats are:"+
```
Comment (kevin-wangzefeng) [inline cmd/controller-manager/app/options/options.go:92]:
```suggestion
	flags.StringVar(&o.SkippedPropagatingAPIs, "skippedPropagatingAPIs", "", "Semicolon separated resources that should be skipped from propagating in addition to the default skip list(cluster.karmada.io;policy.karmada.io;work.karmada.io). Supported formats are:"+
```

File: pkg/util/apigroup.go [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+	// group version kind networking.k8s.io/v1/Ingress
+	// corev1 has no group
+	tokens := strings.Split(c, ";")
+	for _, token := range tokens {
+		r.parseSingle(token)
+	}
+}
+
+func (r *ResourceConfig) parseSingle(token string) {
+	var groupVersionKindPattern = regexp.MustCompile(`^([\w.]+)/(\w+)/([\w,]*)`)
```
Comment (kevin-wangzefeng) [inline pkg/util/apigroup.go:47]: Not quite sure how `[.]` is parsed in the golang regexp, do we need escape with `\`? I mean `[\w\.]`

File: pkg/util/detector/detector.go [hunk 6]
```diff
@@ -199,6 +200,11 @@ func (d *ResourceDetector) Reconcile(key util.QueueKey) error {
 // All objects which API group defined by Karmada should be ignored:
 // - cluster.karmada.io
 // - policy.karmada.io
+//
+//
+// if skippedPropagatingAPIs is set, Karmada will not ignore these apis.
```
Comment (kevin-wangzefeng) [inline pkg/util/detector/detector.go:205]: The api objects listed above need to be ignored anyway (or leasted by default), we don't want users to manually input the things they don't care when trying to skip something else.

File: cmd/controller-manager/app/options/options.go [hunk 7]
```diff
@@ -103,4 +105,8 @@ func (o *Options) AddFlags(flags *pflag.FlagSet) {
 		"Specifies the grace period of allowing a running cluster to be unresponsive before marking it unhealthy.")
 	flags.DurationVar(&o.ClusterStartupGracePeriod.Duration, "cluster-startup-grace-period", 60*time.Second,
 		"Specifies the grace period of allowing a cluster to be unresponsive during startup before marking it unhealthy.")
+	flags.StringVar(&o.SkippedPropagatingAPIs, "skippedPropagatingAPIs", "", "Semicolon separated resources that should be skipped from propagating in addition to the default skip list(cluster.karmada.io;policy.karmada.io;work.karmada.io). Supported formats are:"+
+		"<group> for skip resources with a specific API group(e.g. networking.k8s.io),"+
+		"<group>/<version> for skip resources with a specific API version(e.g. networking.k8s.io/v1beta1),"+
+		"<group>/<version>/<kind>,<kind> for skip one or more specific resource(e.g. networking.k8s.io/v1beta1/Ingress,IngressClass) where the kinds are case-insensitive.")
```
Comment (RainbowMango) [inline cmd/controller-manager/app/options/options.go:111]:
```suggestion
	flags.StringVar(&o.SkippedPropagatingAPIs, "skipped-propagating-apis", "", "Semicolon separated resources that should be skipped from propagating in addition to the default skip list(cluster.karmada.io;policy.karmada.io;work.karmada.io). Supported formats are: \n"+
		"<group> for skip resources with a specific API group(e.g. networking.k8s.io),\n"+
		"<group>/<version> for skip resources with a specific API version(e.g. networking.k8s.io/v1beta1),\n"+
		"<group>/<version>/<kind>,<kind> for skip one or more specific resource(e.g. networking.k8s.io/v1beta1/Ingress,IngressClass) where the kinds are case-insensitive.")
```

Comment (RainbowMango) [reply cmd/controller-manager/app/options/options.go:111]: I tested on my side, the flag looks like:
```
      --skipped-propagating-apis string               Semicolon separated resources that should be skipped from propagating in addition to the default skip list(cluster.karmada.io;policy.karmada.io;work.karmada.io). Supported formats are: 
                                                      <group> for skip resources with a specific API group(e.g. networking.k8s.io),
                                                      <group>/<version> for skip resources with a specific API version(e.g. networking.k8s.io/v1beta1),
                                                      <group>/<version>/<kind>,<kind> for skip one or more specific resource(e.g. networking.k8s.io/v1beta1/Ingress,IngressClass) where the kinds are case-insensitive.
```

File: pkg/controllers/binding/binding_controller.go [hunk 8]
```diff
@@ -77,7 +77,6 @@ func (c *ResourceBindingController) syncBinding(binding *workv1alpha1.ResourceBi
 			binding.GetNamespace(), binding.GetName(), err)
 		return controllerruntime.Result{Requeue: true}, err
 	}
-
```
Comment (RainbowMango) [inline pkg/controllers/binding/binding_controller.go:80]: You remove this emptly line on purpose or made by some tools?

File: pkg/controllers/namespace/namespace_sync_controller.go [hunk 9]
```diff
@@ -10,6 +10,7 @@ import (
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/types"
+
```
Comment (RainbowMango) [inline pkg/controllers/namespace/namespace_sync_controller.go:13]: Why separated them?

File: cmd/controller-manager/app/controllermanager.go [hunk 10]
```diff
@@ -100,17 +100,19 @@ func setupControllers(mgr controllerruntime.Manager, opts *options.Options, stop
 	resetConfig := mgr.GetConfig()
 	dynamicClientSet := dynamic.NewForConfigOrDie(resetConfig)
 	discoverClientSet := discovery.NewDiscoveryClientForConfigOrDie(resetConfig)
-
 	objectWatcher := objectwatcher.NewObjectWatcher(mgr.GetClient(), mgr.GetRESTMapper(), util.NewClusterDynamicClientSet)
 	overrideManager := overridemanager.New(mgr.GetClient())
-
+	resourceConfig := util.NewResourceConfig()
+	resourceConfig.Parse(opts.SkippedPropagatingAPIs)
```
Comment (RainbowMango) [inline cmd/controller-manager/app/controllermanager.go:106]: What if the input format is wrong? We should dump the error message and stop.

File: pkg/util/apigroup.go (hunk 5 above)
Comment (RainbowMango) [reply pkg/util/apigroup.go:47]: Regexp is a little bit complicated and hard to understand.
Can we parse it like the [ParseGroupVersion](https://github.com/kubernetes/kubernetes/blob/696d0f5772d030a37ea67056b23e75330321a345/staging/src/k8s.io/apimachinery/pkg/runtime/schema/group_version.go#L211)?

Comment (RainbowMango) [reply pkg/util/apigroup.go:47]: We need a unit test here by the way.

File: pkg/util/detector/detector.go [hunk 11]
```diff
@@ -13,6 +13,8 @@ import (
 	appsv1 "k8s.io/api/apps/v1"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
```
Comment (RainbowMango) [inline pkg/util/detector/detector.go:17]: Are you using some weird automation tools?

File: pkg/util/detector/detector.go [hunk 12]
```diff
@@ -199,6 +200,14 @@ func (d *ResourceDetector) Reconcile(key util.QueueKey) error {
 // All objects which API group defined by Karmada should be ignored:
 // - cluster.karmada.io
 // - policy.karmada.io
+//
+// The api objects listed above need to be ignored anyway
+// (or leasted by default), we don't want users to manually input the things
+// they don't care when trying to skip something else.
```
Comment (RainbowMango) [inline pkg/util/detector/detector.go:206]:
```suggestion
// The api objects listed above will be ignored by default, as we don't want users to manually input the things
// they don't care when trying to skip something else.
```

File: pkg/util/detector/detector.go [hunk 13]
```diff
@@ -199,6 +200,14 @@ func (d *ResourceDetector) Reconcile(key util.QueueKey) error {
...
+//
+// if skippedPropagatingAPIs is set, Karmada will not ignore these apis.
+//
+//
```
Comment (RainbowMango) [inline pkg/util/detector/detector.go:210]:
```suggestion
// If '--skipped-propagating-apis' which used to specific the APIs should be ignored in addition to the defaults, is set,
// the specified apis will be ignored as well.
```

File: pkg/util/detector/detector.go [hunk 14]
```diff
@@ -216,12 +225,27 @@ func (d *ResourceDetector) EventFilter(obj interface{}) bool {
...
+			return false
+		}
+		if d.ResourceConfig.GroupVersionDisabled(gv) {
+			klog.V(4).Infof("Skip event for %s", clusterWideKey)
+			return false
+		}
+		if d.ResourceConfig.GroupVersionKindDisabled(gvk) {
+			klog.V(4).Infof("Skip event for %s", clusterWideKey)
+			return false
+		}
```
Comment (RainbowMango) [inline pkg/util/detector/detector.go:247]: I guess we didn't make an agreement about this flag.

The flag used to specify `extra` APIs should be ignored, not used to `replace` the defaults.

File: cmd/controller-manager/app/controllermanager.go (hunk 10 above)
Comment (pigletfly) [reply cmd/controller-manager/app/controllermanager.go:106]: yes, We should validate first

File: pkg/util/detector/detector.go (hunk 14 above)
Comment (pigletfly) [reply pkg/util/detector/detector.go:247]: yes, I moved these default into 
```
	// disable karmada group by default
	r.DisableGroup(clusterv1alpha1.GroupVersion.Group)
	r.DisableGroup(policyv1alpha1.GroupVersion.Group)
	r.DisableGroup(workv1alpha1.GroupVersion.Group)
```

File: pkg/util/apigroup.go (hunk 5 above)
Comment (pigletfly) [reply pkg/util/apigroup.go:47]: I will dig into it.Regex is complicated indeed.

File: cmd/controller-manager/app/controllermanager.go (hunk 10 above)
Comment (pigletfly) [reply cmd/controller-manager/app/controllermanager.go:107]: Add a TODO

**Comment by pigletfly** [comment]
@RainbowMango PTAL

**Comment by RainbowMango** [comment]
OK. I'll do it later today.

File: pkg/util/detector/detector.go [hunk 15]
```diff
@@ -217,12 +224,27 @@ func (d *ResourceDetector) EventFilter(obj interface{}) bool {
...
+		gvk := schema.GroupVersionKind{
+			Group:   clusterWideKey.Group,
+			Version: clusterWideKey.Version,
+			Kind:    clusterWideKey.Kind,
+		}
+		if d.ResourceConfig.GroupDisabled(clusterWideKey.Group) {
+			klog.V(4).Infof("Skip event for %s", clusterWideKey)
+			return false
+		}
+		if d.ResourceConfig.GroupVersionDisabled(gv) {
```
Comment (RainbowMango) [inline pkg/util/detector/detector.go:239]:
```suggestion
		if d.ResourceConfig.GroupVersionDisabled(clusterWideKey.GroupVersion()) {
```

File: pkg/util/detector/detector.go [hunk 16]
```diff
@@ -217,12 +224,27 @@ func (d *ResourceDetector) EventFilter(obj interface{}) bool {
...
+			klog.V(4).Infof("Skip event for %s", clusterWideKey)
+			return false
+		}
+		if d.ResourceConfig.GroupVersionKindDisabled(gvk) {
```
Comment (RainbowMango) [inline pkg/util/detector/detector.go:243]:
```suggestion
		if d.ResourceConfig.GroupVersionKindDisabled(clusterWideKey.GroupVersionKind()) {
```

**Review by RainbowMango — COMMENTED** [review]
Just some nits, otherwise LGTM.

Just a reminder, we are not using a lock to protect the maps between read/write operations.
I don't know if it is done on purpose because it'll works fine since the write operations only happens at startup phase.

**Comment by pigletfly** [comment]
@RainbowMango PTAL

**Comment by RainbowMango** [comment]
/lgtm
Wait for the CI result.

**Comment by RainbowMango** [comment]
Don't know why the e2e test keeps failing.
@pigletfly Can you rebase your code? I've increased the E2E timeout yesterday by #491 .

```
I0708 01:37:21.460302   28633 propagationpolicy_test.go:195] Waiting for service(karmadatest-rkm/service-w6k) present on cluster(member1)
STEP: removing policy(karmadatest-rkm/service-w6k)

• Failure [60.038 seconds]
[BasicPropagation] basic propagation testing
/home/runner/work/karmada/karmada/test/e2e/propagationpolicy_test.go:28
  Service propagation testing
  /home/runner/work/karmada/karmada/test/e2e/propagationpolicy_test.go:151
    service propagation testing [It]
    /home/runner/work/karmada/karmada/test/e2e/propagationpolicy_test.go:184

    Unexpected error:
        <*errors.errorString | 0xc000200230>: {
            s: "timed out waiting for the condition",
        }
        timed out waiting for the condition
    occurred
```

**Comment by pigletfly** [comment]
rebased

**Review by RainbowMango — APPROVED** [review]

**Pull request merged**